File: backend/src/api/routes/shows.ts
import { Router, Response } from 'express';
import { AuthRequest, authMiddleware } from '../middleware/auth';
import { prisma } from '../../lib/db';
import { createPodcastSchema, updatePodcastSchema } from '../../lib/validators';
import { ApiError, NotFoundError, ForbiddenError, handleError } from '../../lib/errors';
import { logger } from '../../lib/logger';
import { checkPodcastOwnership, requirePodcastAccess } from '../../lib/authorization';

const router = Router();

// Middleware - require authentication
router.use(authMiddleware);

// Create a new podcast
router.post('/', async (req: AuthRequest, res: Response) => {
  try {
    const validationResult = createPodcastSchema.safeParse(req.body);
    if (!validationResult.success) {
      return res.status(400).json({ error: 'Invalid input', details: validationResult.error });
    }

    const data = validationResult.data;
    const podcast = await prisma.podcast.create({
      data: {
        ...data,
        ownerId: req.userId!,
      },
    });

    logger.info('Podcast created', { podcastId: podcast.id, userId: req.userId });
    res.status(201).json(podcast);
  } catch (error) {
    const { statusCode, body } = handleError(error);
    res.status(statusCode).json(body);
  }
});

// Get all podcasts for current user
router.get('/', async (req: AuthRequest, res: Response) => {
  try {
    const podcasts = await prisma.podcast.findMany({
      where: {
        OR: [
          { ownerId: req.userId },
          {
            teamMembers: {
              some: { userId: req.userId },
            },
          },
        ],
      },
      include: {
        artwork: true,
        _count: { select: { episodes: true } },
      },
    });

    res.json(podcasts);
  } catch (error) {
    const { statusCode, body } = handleError(error);
    res.status(statusCode).json(body);
  }
});

// Get a specific podcast
router.get('/:id', async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;

    // Check access
    const hasAccess = await requirePodcastAccess()(req.userId!, id);
    if (!hasAccess) {
      throw new ForbiddenError('You do not have access to this podcast');
    }

    const podcast = await prisma.podcast.findUnique({
      where: { id },
      include: {
        artwork: true,
        episodes: {
          select: {
            id: true,
            title: true,
            publishedAt: true,
            episodeNumber: true,
            seasonNumber: true,
          },
        },
        teamMembers: {
          select: {
            user: { select: { id: true, name: true, email: true } },
            role: true,
          },
        },
      },
    });

    if (!podcast) {
      throw new NotFoundError('Podcast');
    }

    res.json(podcast);
  } catch (error) {
    const { statusCode, body } = handleError(error);
    res.status(statusCode).json(body);
  }
});

// Update a podcast
router.put('/:id', async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;

    // Check ownership
    const isOwner = await checkPodcastOwnership(req.userId!, id);
    if (!isOwner) {
      throw new ForbiddenError('Only the podcast owner can update it');
    }

    const validationResult = updatePodcastSchema.safeParse(req.body);
    if (!validationResult.success) {
      return res.status(400).json({ error: 'Invalid input', details: validationResult.error });
    }

    const podcast = await prisma.podcast.update({
      where: { id },
      data: validationResult.data,
    });

    logger.info('Podcast updated', { podcastId: id, userId: req.userId });
    res.json(podcast);
  } catch (error) {
    const { statusCode, body } = handleError(error);
    res.status(statusCode).json(body);
  }
});

// Delete a podcast
router.delete('/:id', async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;

    // Check ownership
    const isOwner = await checkPodcastOwnership(req.userId!, id);
    if (!isOwner) {
      throw new ForbiddenError('Only the podcast owner can delete it');
    }

    await prisma.podcast.delete({ where: { id } });

    logger.info('Podcast deleted', { podcastId: id, userId: req.userId });
    res.status(204).send();
  } catch (error) {
    const { statusCode, body } = handleError(error);
    res.status(statusCode).json(body);
  }
});

// Get podcast feed URL
router.get('/:id/feed-url', async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;

    // Check access
    const hasAccess = await requirePodcastAccess()(req.userId!, id);
    if (!hasAccess) {
      throw new ForbiddenError('You do not have access to this podcast');
    }

    const podcast = await prisma.podcast.findUnique({
      where: { id },
      select: { feedUrl: true },
    });

    if (!podcast) {
      throw new NotFoundError('Podcast');
    }

    const feedUrl = `${process.env.API_HOST || 'http://localhost:3001'}/feeds/shows/${podcast.feedUrl}/rss.xml`;

    res.json({ feedUrl });
  } catch (error) {
    const { statusCode, body } = handleError(error);
    res.status(statusCode).json(body);
  }
});

export default router;

File: backend/app/api/routes/shows.py
"""Podcast shows management endpoints."""

from uuid import uuid4

from fastapi import APIRouter, Depends, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.middleware.auth import get_current_user
from app.core.database import get_db
from app.core.errors import NotFoundException, UnauthorizedException
from app.lib.validators import (
    PodcastCreate,
    PodcastResponse,
    PodcastUpdate,
)
from app.models.base import Podcast, User

router = APIRouter()


@router.post("/", response_model=PodcastResponse, status_code=status.HTTP_201_CREATED)
async def create_podcast(
    podcast_data: PodcastCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PodcastResponse:
    """
    Create a new podcast.

    - **title**: Podcast title
    - **description**: Podcast description
    - **language**: Language code (e.g., 'en', 'ja')
    """
    # Generate unique feed URL
    feed_url = f"podcast-{uuid4().hex[:8]}"

    # Create new podcast
    new_podcast = Podcast(
        id=str(uuid4()),
        owner_id=current_user.id,
        title=podcast_data.title,
        description=podcast_data.description,
        author=podcast_data.author or current_user.full_name,
        category=podcast_data.category,
        language=podcast_data.language,
        feed_url=feed_url,
        artwork_url=None,
        is_published=False,
    )

    db.add(new_podcast)
    await db.commit()
    await db.refresh(new_podcast)

    return PodcastResponse.from_orm(new_podcast)


@router.get("/{podcast_id}", response_model=PodcastResponse)
async def get_podcast(
    podcast_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PodcastResponse:
    """
    Get podcast details by ID.
    """
    result = await db.execute(select(Podcast).where(Podcast.id == podcast_id))
    podcast = result.scalar_one_or_none()

    if not podcast:
        raise NotFoundException(f"Podcast '{podcast_id}' not found")

    return PodcastResponse.from_orm(podcast)


@router.get("", response_model=list[PodcastResponse])
async def list_podcasts(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[PodcastResponse]:
    """
    List all podcasts owned by the current user.
    """
    result = await db.execute(select(Podcast).where(Podcast.owner_id == current_user.id))
    podcasts = result.scalars().all()

    return [PodcastResponse.from_orm(p) for p in podcasts]


@router.put("/{podcast_id}", response_model=PodcastResponse)
async def update_podcast(
    podcast_id: str,
    podcast_data: PodcastUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PodcastResponse:
    """
    Update podcast metadata.

    Only podcast owner can update.
    """
    # Fetch podcast
    result = await db.execute(select(Podcast).where(Podcast.id == podcast_id))
    podcast = result.scalar_one_or_none()

    if not podcast:
        raise NotFoundException(f"Podcast '{podcast_id}' not found")

    # Check ownership
    if podcast.owner_id != current_user.id:
        raise UnauthorizedException("Only podcast owner can update podcast")

    # Update fields
    if podcast_data.title is not None:
        podcast.title = podcast_data.title
    if podcast_data.description is not None:
        podcast.description = podcast_data.description
    if podcast_data.author is not None:
        podcast.author = podcast_data.author
    if podcast_data.category is not None:
        podcast.category = podcast_data.category
    if podcast_data.language is not None:
        podcast.language = podcast_data.language
    if podcast_data.artwork_url is not None:
        podcast.artwork_url = podcast_data.artwork_url

    await db.commit()
    await db.refresh(podcast)

    return PodcastResponse.from_orm(podcast)


@router.delete("/{podcast_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_podcast(
    podcast_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """
    Delete a podcast (owner only).

    Also deletes all related episodes, audio files, and artwork.
    """
    # Fetch podcast
    result = await db.execute(select(Podcast).where(Podcast.id == podcast_id))
    podcast = result.scalar_one_or_none()

    if not podcast:
        raise NotFoundException(f"Podcast '{podcast_id}' not found")

    # Check ownership
    if podcast.owner_id != current_user.id:
        raise UnauthorizedException("Only podcast owner can delete podcast")

    # Delete podcast (cascade deletes related records)
    await db.delete(podcast)
    await db.commit()
